src/Bedrock.Framework/Protocols/IMessageWriter.cs

```diff
@@ -4,6 +4,6 @@ namespace Bedrock.Framework.Protocols
 {
     public interface IMessageWriter<TMessage>
     {
-        void WriteMessage(TMessage message, IBufferWriter<byte> output);
+        void WriteMessage(ref TMessage message, IBufferWriter<byte> output);
```

Comment:
This may be a controversial change, but possibly worth discussing.

For messages that are structs (for avoiding allocations), there may be cases where the message writer may be updating state in the message itself. However, unless this a ref param, any state changes to the message the writer makes are going to be lost.

Side-effect free is probably more ideal and less dangerous, of course. The reason I added this here, in my case, is because the `WebSocketWriteFrame` needs to track whether or not the underlying sequence has already been masked by the writer. There are a couple of alternatives to this change:

* Make `WebSocketWriteFrame` a class and live with the allocations, while still keeping the masking state of the message in the class.
* Keep `WebSocketWriteFrame` a struct and copy the data while masking to a buffer from a pool, and unmask the copy (this is the approach used in corefx).

Reply:
It might be better to have a reusable class then pass a mutable struct around.

Reply:
Going to roll this back for now. It feels a bit dirty overall.